### PR TITLE
PR for Workshop 3 by Rakesh Kumar Singh

### PR DESCRIPTION
**Pull request by rakeshkumar1019**

### CODE REVIEW

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 1]
```diff
@@ -84,13 +92,29 @@ suite('nuxeo-favorites-toggle-button', () => {
...
+    teardown(() => {
+      if (window.confirm.restore) {
+        window.confirm.restore();
+      }
+    });
+
+    test('toggle should not remove document from favorites when clicked Cancel', async () => {
+      sinon.stub(window, 'confirm').returns(false);
+      expect(element.favorite).to.be.true;
+    });
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: This test assumes that the document is in favorites. Therefore, it should be in the other test suite "when a document is a favorite"

**Review by Gabez0r — COMMENTED** [review]
Good job. You should consider squashing some commits.

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 2]
```diff
@@ -48,16 +49,29 @@ suite('nuxeo-favorites-toggle-button', () => {
...
-    test('toggle should remove document from favorites', async () => {
+    test('toggle should remove document from favorites when clicked ok', async () => {
       // Remove document from favorites by toggling
+      sinon.stub(window, 'confirm').returns(true);
       tap(element);
       await waitChanged(element, 'favorite');
       expect(element.favorite).to.be.false;
     });
+
+    test('toggle should not remove document from favorites when clicked Cancel', async (done) => {
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: There's no need use the done here and call it later. This is usually used in sync functions.

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 3]
```diff
@@ -84,13 +98,25 @@ suite('nuxeo-favorites-toggle-button', () => {
         '{"entity-type": "document","uid": "1"}',
       ]);
     });
+    teardown(() => {
+      if (window.confirm.restore) {
+        window.confirm.restore();
+      }
+    });
+
+    test('toggle should not add the document to favorites when clicked on Cancel', async (done) => {
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: Same here.

**Review by Gabez0r — COMMENTED** [review]
Looking good.

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 4]
```diff
@@ -48,16 +49,29 @@ suite('nuxeo-favorites-toggle-button', () => {
...
       // Remove document from favorites by toggling
+      sinon.stub(window, 'confirm').returns(true);
       tap(element);
       await waitChanged(element, 'favorite');
       expect(element.favorite).to.be.false;
     });
+
+    test('toggle should not remove document from favorites when clicked Cancel', async (done) => {
+      sinon.stub(window, 'confirm').returns(false);
+      expect(element.favorite).to.be.true;
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js:77]: We didn't actually click the button, so we are not testing anything here.

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 5]
```diff
@@ -84,13 +98,25 @@ suite('nuxeo-favorites-toggle-button', () => {
...
     });
+    teardown(() => {
+      if (window.confirm.restore) {
+        window.confirm.restore();
+      }
+    });
+
+    test('toggle should not add the document to favorites when clicked on Cancel', async (done) => {
+      sinon.stub(window, 'confirm').returns(false);
+      expect(element.favorite).to.be.false;
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: We didn't actually click the button, so we are not testing anything here.

File: ui/actions/nuxeo-favorites-toggle-button.js [hunk 6]
```diff
@@ -147,25 +147,29 @@ import '../nuxeo-button-styles.js';
...
         this.$.opAdd.execute().then(() => {
-          this.dispatchEvent(
-            new CustomEvent('added-to-favorites', {
-              composed: true,
-              bubbles: true,
-              detail: { doc: this.document },
-            }),
-          );
-          this._setFavorite(true);
+          if (window.confirm(this.i18n('favoritesToggleButton.confirm.add')) === true) {
```
Comment (Gabez0r) [inline ui/actions/nuxeo-favorites-toggle-button.js]: Please note that we are asking the user to confirm AFTER we already executed the operation, which is not what we want here.

**Review by Gabez0r — COMMENTED** [review]
Looking good!

**Review by richardsd — COMMENTED** [review]
Looks good to me. Nice work 🙂

**Pull request closed without merging**